ironic-lib/0002-Base-classes-and-examples.patch

```diff
@@ -0,0 +1,2283 @@
++class SwRaidSetup(Setup):
++    """Software RAID setup implementation"""
++
++    conf_key = 'swraid'
++
++    def get_src_names(self):
++        raid_partitions = [r['partitions'] for r in self.conf.values()]
++        return sum(raid_partitions, [])
++
++    def get_dst_names(self):
++        return list(self.conf)
+diff --git a/ironic_lib/system_installer/systemsetup.py b/ironic_lib/system_installer/systemsetup.py
+new file mode 100644
+index 0000000..e03c5b4
+--- /dev/null
++++ b/ironic_lib/system_installer/systemsetup.py
+@@ -0,0 +1,20 @@
++# Copyright (c) 2018 Intel Corporation
++#
++# Licensed under the Apache License, Version 2.0 (the "License");
++# you may not use this file except in compliance with the License.
++# You may obtain a copy of the License at
++#
++#    http://www.apache.org/licenses/LICENSE-2.0
++#
++# Unless required by applicable law or agreed to in writing, software
++# distributed under the License is distributed on an "AS IS" BASIS,
++# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
++# implied.
++# See the License for the specific language governing permissions and
++# limitations under the License.
++
++from ironic_lib.system_installer.filesystemsetup import FilesystemSetup
++
++
++class SystemSetup(FilesystemSetup):
++    """Basic disk formatting implementation"""
+diff --git a/ironic_lib/system_installer/tools.py b/ironic_lib/system_installer/tools.py
+new file mode 100644
+index 0000000..0dcc435
+--- /dev/null
++++ b/ironic_lib/system_installer/tools.py
+@@ -0,0 +1,55 @@
++# Copyright (c) 2018 Intel Corporation
++#
++# Licensed under the Apache License, Version 2.0 (the "License");
++# you may not use this file except in compliance with the License.
++# You may obtain a copy of the License at
++#
++#    http://www.apache.org/licenses/LICENSE-2.0
++#
++# Unless required by applicable law or agreed to in writing, software
++# distributed under the License is distributed on an "AS IS" BASIS,
++# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
++# implied.
++# See the License for the specific language governing permissions and
++# limitations under the License.
++
++import collections
++import functools
++import json
++import time
++import yaml
++
++
++def ordered_load(disk_conf):
++    """Load JSON or YAML configuration as OrderedDicts"""
++    try:  # Parse JSON as OrderedDicts
++        return json.loads(disk_conf, object_pairs_hook=collections.OrderedDict)
++    except ValueError:
++        # Parse YAML as OrderedDicts
++        class OrderedLoader(yaml.SafeLoader):
++            pass
++
++        def construct_mapping(loader, node):
++            loader.flatten_mapping(node)
++            return collections.OrderedDict(loader.construct_pairs(node))
++        OrderedLoader.add_constructor(
++            yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG,
++            construct_mapping)
++        return yaml.load(disk_conf, OrderedLoader)
++
++
++def retry(tries, exc=(Exception,)):
```

Comment:
Might be able to just use this instead? https://docs.openstack.org/oslo.service/ocata/api/loopingcall.html#oslo_service.loopingcall.RetryDecorator

Reply:
@PiotrProkop ^

Reply:
sure didn't notice that 😄